pkg/kubernetes/api_eks_client.go

```diff
@@ -0,0 +1,188 @@
+package client
+
+import (
+	"context"
+	"encoding/base64"
+	"fmt"
+	"log/slog"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/config"
+	"github.com/aws/aws-sdk-go-v2/service/eks"
+	smithyhttp "github.com/aws/smithy-go/transport/http"
+	"k8s.io/client-go/rest"
+
+	"github.com/aws/aws-sdk-go-v2/service/sts"
+)
+
+const (
+	clusterIDHeader        = "x-k8s-aws-id"
+	presignedURLExpiration = 10 * time.Minute
+	v1Prefix               = "k8s-aws-v1."
+)
+
+// getEKSToken returns a pre-signed token for accessing the EKS cluster.
+func getEKSToken(ctx context.Context, cfg *aws.Config, clusterName string) (*string, error) {
+	stsClient := sts.NewFromConfig(*cfg)
+	// Create the pre-signed request
+	presignClient := sts.NewPresignClient(stsClient)
+	presignedURLRequest, err := presignClient.PresignGetCallerIdentity(ctx, &sts.GetCallerIdentityInput{},
+		func(presignOptions *sts.PresignOptions) {
+			presignOptions.ClientOptions = append(presignOptions.ClientOptions, appendPresignHeaderValuesFunc(clusterName))
+		})
+	if err != nil {
+		return nil, fmt.Errorf("failed to presign caller identity: %w", err)
+	}
+
+	// Encode the pre-signed URL
+	token := v1Prefix + base64.RawURLEncoding.EncodeToString([]byte(presignedURLRequest.URL))
+	return &token, nil
+}
+
+func appendPresignHeaderValuesFunc(clusterID string) func(stsOptions *sts.Options) {
+	return func(stsOptions *sts.Options) {
+		stsOptions.APIOptions = append(stsOptions.APIOptions,
+			// Add clusterId Header.
+			smithyhttp.SetHeaderValue(clusterIDHeader, clusterID),
+			// Add X-Amz-Expires query param.
+			smithyhttp.SetHeaderValue("X-Amz-Expires", "60"))
+	}
+}
+
+// EKSClientOption sets up a Kubernetes client for accessing an AWS EKS cluster.
+//
+// This function assumes that the AWS SDK configuration can obtain the necessary credentials
+// from the environment. This is the default behavior of the AWS SDK. It is recommended to
+// use IAM Roles for Service Accounts (IRSA) for EKS clusters to grant IAM roles to the pods.
+// Using IRSA provides fine-grained permissions and security best practices. If IRSA is not
+// used, the pod will attempt to use the instance credentials and generate an STS token for
+// accessing the EKS cluster.
+//
+// Parameters:
+//
+//   - ctx: the context for the AWS SDK client.
+//
+//   - clusterID: the name of the EKS cluster.
+//
+//   - region: the AWS region where the EKS cluster is located. (optional) If not provided, the default region will be used.
+func EKSClientOption(ctx context.Context, clusterID, region string) NewClientOption {
```

Comment:
I don't think you need `region`, since `AWS_REGION` will let users override this themselves directly.

Reply:
@djeebus i've removed the region parameter.